lib/spack/spack/cmd/ci.py

```diff
@@ -650,6 +672,166 @@ def _gitlab_artifacts_url(url: str) -> str:
     return urlunparse(parsed._replace(path="/".join(parts), fragment="", query=""))
 
 
+def validate_standard_versions(pkg, versions: VersionList) -> bool:
+    """Get and test the checksum of a package version based on a tarball.
+    Args:
+      pkg (spack.package_base.PackageBase): Spack package for which to validate a version checksum
+      versions (spack.version.VersionList): list of package versions to validate
+    Returns: (bool): result of the validation. True is valid and false is failed.
+    """
+    url_dict: Dict[StandardVersion, str] = {}
+
+    for version in versions:
+        url = pkg.find_valid_url_for_version(version)
+        url_dict[version] = url
+
+    version_hashes = spack.stage.get_checksums_for_versions(
+        url_dict, pkg.name, fetch_options=pkg.fetch_options
+    )
+
+    valid_checksums = True
+    for version, sha in version_hashes.items():
+        if sha != pkg.versions[version]["sha256"]:
+            tty.error(
+                f"Invalid checksum found {pkg.name}@{version}\n"
+                f"    [package.py] {pkg.versions[version]['sha256']}\n"
+                f"    [Downloaded] {sha}"
+            )
+            valid_checksums = False
+            continue
+
+        tty.info(f"Validated {pkg.name}@{version} --> {sha}")
+
+    return valid_checksums
+
+
+def validate_git_versions(pkg, versions: VersionList) -> bool:
+    """Get and test the commit and tag of a package version based on a git repository.
+    Args:
+      pkg (spack.package_base.PackageBase): Spack package for which to validate a version
+      versions (spack.version.VersionList): list of package versions to validate
+    Returns: (bool): result of the validation. True is valid and false is failed.
+    """
+    git = spack.util.git.git(required=True)
+
+    with tempfile.TemporaryDirectory() as tmpdirpath:
+        # Test if repository can be cloned.
+        try:
+            _ = git("clone", pkg.git, tmpdirpath, output=str, error=str)
+        except ProcessError as exp:
+            tty.error(f"Unable to clone git repository for {pkg.name}", exp)
+            return False
+
+        valid_commit = True
+        for version in versions:
+            known_commit = pkg.versions[version]["commit"]
+            with fs.working_dir(tmpdirpath):
+                # Test if the specified commit is in the repository.
+                # If the commit is located in the repository the command will return 0 and
+                # print the word "commit" else it will fail with a lookup error
+                try:
+                    git("cat-file", "-t", known_commit, output=str, error=str)
```

Comment:
if you just clone the repository at a particular commit, `git` does an integrity check, which is all you need here. The fetcher does this, so  think just running the fetcher is all you need to do in most cases.  See [this thread on the git mailing list](https://lore.kernel.org/git/ED97E252-CABA-41BB-B18C-819A5EF305E3@llnl.gov/).  If you use a common repo to validate multiple commits, I *think* you want to check each one out to validate it.

The fetcher by default does a shallow clone -- which I think will save time unless you have a lot of versions to verify.  I *believe* you still get the integrity check for the commit you asked for, but I am not entirely sure, so check on that.

Reply:
Unified with the existing fetch strategies and now check that the repository checks out successfully.